### PR TITLE
Adding finished training script

### PR DESCRIPTION
**Pull request by vasudevgupta7**
This PR adds training script used for getting `5.6% WER` (as discussed in #13). There were lots of bugs earlier, but now everything is fixed and working properly. Also, realised that we don't need custom training loop, hence removed that.

@sayakpaul @MorganR kindly review the PR

Thanks!!

### CODE REVIEW

File: src/main.py
Comment (sayakpaul) [inline src/main.py:61]: What will these buckets contain? Add a small comment to explain that.

File: src/main.py
Comment (sayakpaul) [inline src/main.py:68]: What will these buckets contain? Add a small comment to explain that.

File: src/main.py
Comment (sayakpaul) [inline src/main.py:75]: Same.

File: src/main.py
Comment (sayakpaul) [inline src/main.py:201]: A better and more informative comment might be: 

> Freezing the backbone.

Or the section of the network we are freezing like stem, body, trunk, etc.

File: src/main.py
Comment (sayakpaul) [inline src/main.py:232]: I think we can remove this print statement as it'd make the logs verbose.

Comment (sayakpaul) [inline src/main.py:232]: I am a bit confused about this. During fine-tuning, we actually unfreeze the middle layers, pre-final layers and then continue the process with smaller learning rates so that the pre-trained weights do not break too fast. But what's the notion here? 

Also, since you're already doing `model.layers[0].trainable = False` then by default, all the layers under it are getting non-trainable.  Know more about this here: https://keras.io/guides/transfer_learning/.

**Review by sayakpaul — CHANGES_REQUESTED** [review]

File: src/main.py
Comment (vasudevgupta7) [reply src/main.py:232]: In speech models like Wav2Vec2, it's recommened to keep feature extraction layers (initial convolutional layers) freezed throughout the fine-tuning process. For fine-tuning also, we just need to train all transformer blocks & LM layer (top most layer).

Here training is divided into 2 stages:
1. freeze the pre-trained model and only train the LM head until model converges
2. unfreeze complete model (keep conv feature extraction layer freezed) and train the model

`model.layers[0].trainable = False` is helping in first stage but for 2nd stage we are unfreezing complete model again using `model.trainable = True`.

Comment (sayakpaul) [reply src/main.py:232]: Clears my confusion. We might want to add extensive comments in this part.

**Comment by vasudevgupta7** [comment]
@sayakpaul @MorganR 

I have tried to improve README a bit. Please give your feedback on that as well.

Other than that, I made a small demo showing how good wav2vec2 model is. Here is the link: https://huggingface.co/spaces/vasudevgupta/GOOGLE_SUMMER_OF_CODE. It's deployed in real time with HuggingFace spaces & gradio. I am going to improve that a bit in near future.

**Comment by sayakpaul** [comment]
@vasudevgupta7 this is neat stuff. Just a question out of curiosity. Do you plan to also open-source the code/make a note on preparing that interface on Spaces in the repository?

**Comment by vasudevgupta7** [comment]
Code is already open sourced on HuggingFace Hub ([see this](https://huggingface.co/spaces/vasudevgupta/GOOGLE_SUMMER_OF_CODE/tree/main)). It's not like actual deployed model (in which people do lots of cool things) and made mostly for small demo purposes.

**Comment by vasudevgupta7** [comment]
@MorganR, looking forward to feedback from you as well. I will merge once you approve.

Thanks!!

**Comment by vasudevgupta7** [comment]
Merging this PR as I need `main` branch to be updated for making final report. But would be happy to address any issue in separate PR if @MorganR has suggestions regarding this PR.

**Pull request merged**
